### ISSUE TITLE
converting html tables to markdown creates unusable tables

### ISSUE THREAD

**Issue by osmanburucu-ibm** [didix21/mdutils#73]
**Describe the bug**
cnverting from a wordpress based html content to markdown the tables are broken

**To Reproduce**
Based on this page: https://www.urbancode.com/plugindoc/accurev-scm#tab-settings
original HTML part which is converted:
<div class="section-title">
<h3>Steps and Roles</h3>
</div>
<div class="pn-col-6-6">
<p></p>
<p></p>
<h2><a name="steps"></a>Process steps in the AccuRev plug-in</h2>
<ul>
<li><a href="#import_version">Import Version</a></li>
</ul>
<p></p>
<p></p>
<h3><a name="import_version"></a>Import Version</h3>
<p>Creates a new component version and imports artifacts from AccuRev.</p>
<p>This step has no input properties.</p>
<p></p>
<h2><a name="roles"></a>Roles in the AccuRev plug-in</h2>
<p>The plug-in adds these roles automatically to resources. You cannot add these roles<br/>
   manually.
</p>
<ul>
<li><a href="#accurevcomponentproperties_role">AccuRevComponentProperties</a></li>
<li><a href="#accurevimportproperties_role">AccuRevImportProperties</a></li>
</ul>
<p></p>
<h3><a name="accurevcomponentproperties_role"></a>AccuRevComponentProperties</h3>
<table>
<caption style="text-align: left;"><em>Properties for the AccuRevComponentProperties role</em></caption>
<tbody>
<tr>
<th id="col1">Name</th>
<th id="col2">Type</th>
<th id="col3">Description</th>
</tr>
</tbody>
<tbody>
<tr>
<td headers="col1">AccuRev Executable Path</td>
<td headers="col2">String</td>
<td headers="col3">Full path to the AccuRev executable on the agents file system.</td>
</tr>
<tr>
<td headers="col1">AccuRev URL</td>
<td headers="col2">String</td>
<td headers="col3">Full URL for the AccuRev server. Hostname and port can<br/>
be identified by running the `accurev info` command locally.</td>
</tr>
<tr>
<td headers="col1">Depot</td>
<td headers="col2">String</td>
<td headers="col3">Specify a Depot to retrieve artifacts from.</td>
</tr>
<tr>
<td headers="col1">Excludes</td>
<td headers="col2">String</td>
<td headers="col3">Specify a list of files to exclude during artifact import.</td>
</tr>
<tr>
<td headers="col1">Import Issues</td>
<td headers="col2">Boolean</td>
<td headers="col3">Select to import latest Issue change set artifacts on automatic<br/>
import. Artifacts will still be retrieved if the Issue<br/>
number is specified on manual import. Only valid for a non-Workspace Stream.</td>
</tr>
<tr>
<td headers="col1">Import Promote Transactions</td>
<td headers="col2">Boolean</td>
<td headers="col3">Select to import latest Promote Transaction artifacts on automatic<br/>
import. Artifacts will still be retrieved if the Transaction<br/>
number is specified on manual import. Only valid for a non-Workspace Stream.</td>
</tr>
<tr>
<td headers="col1">Import Snapshots</td>
<td headers="col2">Boolean</td>
<td headers="col3">Select to import latest Snapshot artifacts on automatic<br/>
import. Artifacts will still be retrieved if the Snapshot name<br/>
is specified on manual import. Only valid for a non-Workspace Stream.</td>
</tr>
<tr>
<td headers="col1">Includes</td>
<td headers="col2">String</td>
<td headers="col3">Specify a list of files to include during artifact import.</td>
</tr>
<tr>
<td headers="col1">Password</td>
<td headers="col2">Password</td>
<td headers="col3">Specify the password of the user with permissions to access<br/>
the AccuRev Depot and Stream.</td>
</tr>
<tr>
<td headers="col1">Preserve Execute Permissions</td>
<td headers="col2">Boolean</td>
<td headers="col3">Select to save file execute permissions with files.</td>
</tr>
<tr>
<td headers="col1">Stream</td>
<td headers="col2">String</td>
<td headers="col3">Specify a Stream or Workspace to retrieve artifacts from.<br/>
If empty, the Depot name will be used. If a Workspace is specified, the<br/>
checkboxes and manual import values for Promote Transactions, Issues, and<br/>
Snapshot will be ignored. A version will be created based on the runtime<br/>
timestamp and file contents of the Workspace.</td>
</tr>
<tr>
<td headers="col1">Username</td>
<td headers="col2">String</td>
<td headers="col3">Specify a user with permissions to access the AccuRev Depot and Stream.</td>
</tr>
</tbody>
</table>
<h3><a name="accurevimportproperties_role"></a>AccuRevImportProperties</h3>
<table>
<caption style="text-align: left;"><em>Properties for the AccuRevImportProperties role</em></caption>
<tbody>
<tr>
<th id="col1">Name</th>
<th id="col2">Type</th>
<th id="col3">Description</th>
</tr>
</tbody>
<tbody>
<tr>
<td headers="col1">Issue Numbers</td>
<td headers="col2">String</td>
<td headers="col3">Specify a Issue Number to import. Only valid for a non-Workspace Stream.<br/>
Separate multiple Issue Numbers with commas. Example: 5, 16, 27</td>
</tr>
<tr>
<td headers="col1">Stream Snapshots</td>
<td headers="col2">String</td>
<td headers="col3">Specify a Stream Snapshot to import. Only valid for a non-Workspace Stream.<br/>
Separate multiple Snapshots with a comma. Example: snap-stream1, snap-stream2</td>
</tr>
<tr>
<td headers="col1">Transaction Numbers</td>
<td headers="col2">String</td>
<td headers="col3">Specify a Promote Transaction Number to import. Only valid for a non-Workspace Stream.<br/>
Separate multiple Transaction Numbers with commas. Example: 80, 120, 184</td>
</tr>
<tr>
<td headers="col1">Version Name Tag</td>
<td headers="col2">String</td>
<td headers="col3">Specify a tag to append to the versions name.</td>
</tr>
</tbody>
</table>
<p></p>
</div>

I pre-process the html and remove the following:
<caption style="text-align: left;"><em>Properties for the AccuRevComponentProperties role</em></caption>
<br/> 
-------
This converts to this:

Process steps in the AccuRev plug-in
------------------------------------

* [Import 
Version](#import_version)

### Import Version

Creates a new component version and imports artifacts from AccuRev.

This step has no input properties.

Roles in the AccuRev plug-in
----------------------------

The plug-in adds 
these roles automatically to resources. You cannot add these roles manually.

* 
[AccuRevComponentProperties](#accurevcomponentproperties_role)
* 
[AccuRevImportProperties](#accurevimportproperties_role)

### AccuRevComponentProperties

| Name | Type | 
Description |
| --- | --- | --- |
| AccuRev Executable Path | String | Full path to the AccuRev executable on the agents
 file system. |
| AccuRev URL | String | Full URL for the AccuRev server. Hostname and port can be identified by running
 the `accurev info` command locally. |
| Depot | String | Specify a Depot to retrieve artifacts from. |
| Excludes | 
String | Specify a list of files to exclude during artifact import. |
| Import Issues | Boolean | Select to import 
latest Issue change set artifacts on automatic import. Artifacts will still be retrieved if the Issue number is 
specified on manual import. Only valid for a non-Workspace Stream. |
| Import Promote Transactions | Boolean | Select to
 import latest Promote Transaction artifacts on automatic import. Artifacts will still be retrieved if the Transaction 
number is specified on manual import. Only valid for a non-Workspace Stream. |
| Import Snapshots | Boolean | Select to 
import latest Snapshot artifacts on automatic import. Artifacts will still be retrieved if the Snapshot name is 
specified on manual import. Only valid for a non-Workspace Stream. |
| Includes | String | Specify a list of files to 
include during artifact import. |
| Password | Password | Specify the password of the user with permissions to access 
the AccuRev Depot and Stream. |
| Preserve Execute Permissions | Boolean | Select to save file execute permissions with 
files. |
| Stream | String | Specify a Stream or Workspace to retrieve artifacts from. If empty, the Depot name will be 
used. If a Workspace is specified, the checkboxes and manual import values for Promote Transactions, Issues, and 
Snapshot will be ignored. A version will be created based on the runtime timestamp and file contents of the Workspace. |

| Username | String | Specify a user with permissions to access the AccuRev Depot and Stream. |

### 
AccuRevImportProperties

| Name | Type | Description |
| --- | --- | --- |
| Issue Numbers | String | Specify a Issue
 Number to import. Only valid for a non-Workspace Stream. Separate multiple Issue Numbers with commas. Example: 5, 16, 
27 |
| Stream Snapshots | String | Specify a Stream Snapshot to import. Only valid for a non-Workspace Stream. Separate 
multiple Snapshots with a comma. Example: snap-stream1, snap-stream2 |
| Transaction Numbers | String | Specify a 
Promote Transaction Number to import. Only valid for a non-Workspace Stream. Separate multiple Transaction Numbers with 
commas. Example: 80, 120, 184 |
| Version Name Tag | String | Specify a tag to append to the versions name. |

**Comment by didix21** [comment]
So this output:
```
| Name | Type |
Description |
| --- | --- | --- |
| AccuRev Executable Path | String | Full path to the AccuRev executable on the agents
file system. |
| AccuRev URL | String | Full URL for the AccuRev server. Hostname and port can be identified by running
the accurev info command locally. |
| Depot | String | Specify a Depot to retrieve artifacts from. |
| Excludes |
String | Specify a list of files to exclude during artifact import. |
| Import Issues | Boolean | Select to import
latest Issue change set artifacts on automatic import. Artifacts will still be retrieved if the Issue number is
specified on manual import. Only valid for a non-Workspace Stream. |
| Import Promote Transactions | Boolean | Select to
import latest Promote Transaction artifacts on automatic import. Artifacts will still be retrieved if the Transaction
number is specified on manual import. Only valid for a non-Workspace Stream. |
| Import Snapshots | Boolean | Select to
import latest Snapshot artifacts on automatic import. Artifacts will still be retrieved if the Snapshot name is
specified on manual import. Only valid for a non-Workspace Stream. |
| Includes | String | Specify a list of files to
include during artifact import. |
| Password | Password | Specify the password of the user with permissions to access
the AccuRev Depot and Stream. |
| Preserve Execute Permissions | Boolean | Select to save file execute permissions with
files. |
| Stream | String | Specify a Stream or Workspace to retrieve artifacts from. If empty, the Depot name will be
used. If a Workspace is specified, the checkboxes and manual import values for Promote Transactions, Issues, and
Snapshot will be ignored. A version will be created based on the runtime timestamp and file contents of the Workspace. |

| Username | String | Specify a user with permissions to access the AccuRev Depot and Stream. |
```

is what you find on the markdown file once is generated with mdutils?

**Comment by osmanburucu-ibm** [comment]
Yes this is correct. I can provide you with the raw data if you like. The strange breaks in the line structure can be also seen on non-table paragraphs but as the markdown file is rendered then in GH Pages only the tables are seen as a mess.
here are two examplares of the raw data : 
* https://raw.githubusercontent.com/UrbanCode/IBM-UCx-PLUGIN-DOCS/main/docs/UCD/AccuRevSourceConfig/steps%20and%20roles.md
* https://raw.githubusercontent.com/UrbanCode/IBM-UCx-PLUGIN-DOCS/main/docs/UCD/AccuRevSourceConfig/overview.md

**Comment by didix21** [comment]
The problem probably is due to `wrap_width=120` from  method new_paragraph(self, text='', bold_italics_code='', color='black', align='', wrap_width=120)`. Seems this is generating a lot of problems so maybe I would disable by default.

**Comment by osmanburucu-ibm** [comment]
Just another issue regarding tables:
if a caption style is in the html table, it also breaks the output. At the moment I delete the caption line from the html response before i convert it. Do not have the MD output available at the moment. It adds a **Repository properties** right before the table without any seperation. Maybe you could also check that? **THANK YOU for your response!**

 Here is an sample of the HTML table->
<table>
<caption style="text-align: left;"><em>Repository properties</em></caption>
<tbody>
<tr>
<th id="col1">Name</th>
<th id="col2">Type</th>
<th id="col3">Description</th>
</tr>
</tbody>
<tbody>
<tr>
<td headers="col1">Command Path</td>
<td headers="col2">String</td>
<td headers="col3">The location of the AccuRev command executable file, if the location is not specified on the system PATH statement.
         </td>
</tr>
<tr>
<td headers="col1">Depot Name</td>
<td headers="col2">String</td>
<td headers="col3">The name or ID of the AccuRev depot. The depot is the repository on the server for all related source code. The value for this property can be used as part of the source  config stream name with <strong>${source/repo/depotName}</strong>. If set, this will be recorded with any source changes for use by the AccuWork plug-in.
         </td>
</tr>
<tr>
<td headers="col1">Password</td>
<td headers="col2">Password</td>
<td headers="col3">The password to be used to connect to the AccuRev server.</td>
</tr>
<tr>
<td headers="col1">Password Script</td>
<td headers="col2">String</td>
<td headers="col3">The property to use for the password to connect to the AccuRev server. For example: ${p:AccuRevPassword}. If this property is specified, do not specify a value for the   Password property.
         </td>
</tr>
<tr>
<td headers="col1">Repository Host</td>
<td headers="col2">String</td>
<td headers="col3">The location of the AccuRev server. Specify in the format: <code>host:port</code>.</td>
</tr>
<tr>
<td headers="col1">User Name</td>
<td headers="col2">String</td>
<td headers="col3">The user name to be used to connect to the AccuRev server.</td>
</tr>
</tbody>
</table>
<table>
<caption style="text-align: left;"><em>Integration properties</em></caption>
<tbody>
<tr>
<th id="col1">Name</th>
<th id="col2">Type</th>
<th id="col3">Description</th>
</tr>
</tbody>
<tbody>
<tr>
<td headers="col1">Directory Offset</td>
<td headers="col2">String</td>
<td headers="col3">The directory offset of the working directory for the current job. Use a period (.) to clone into the current working directory.
         </td>
</tr>
<tr>
<td headers="col1">Exclude Filters</td>
<td headers="col2">String</td>
<td headers="col3">A list of user names to exclude.</td>
</tr>
<tr>
<td headers="col1">File Filters</td>
<td headers="col2">String</td>
<td headers="col3">Restrict the changelog to files matching these patterns. Start each lines with a plus sign (+) to include or minus sign (- )to exclude. You can use the following wildcards in the pattern: *, **, and ?.
         </td>
</tr>
<tr>
<td headers="col1">Populate Locations</td>
<td headers="col2">String</td>
<td headers="col3">A list of depot-relative path for the files that you want to populate the workspace or stream. The pop command is use to restore the listed files.  Only the specified files are updated, without building the associated directory tree.
         </td>
</tr>
<tr>
<td headers="col1">Repository</td>
<td headers="col2"></td>
<td headers="col3">The AccuRev repository to be used.</td>
</tr>
<tr>
<td headers="col1">Revision</td>
<td headers="col2">String</td>
<td headers="col3">The version that is to be checked out.</td>
</tr>
<tr>
<td headers="col1">Stream</td>
<td headers="col2">String</td>
<td headers="col3">The stream that is to be checked out.</td>
</tr>
</tbody>
</table>

**Comment by didix21** [comment]
Which methods are you calling? `new_header` and `new_paragraph`?

**Comment by osmanburucu-ibm** [comment]
I first "cleanse the content" so that the tables look as good as possible. 
`    contents = "".join(str(item) for item in soupcontent.contents)
    logger1.info(f"contents-original={contents}")
    # the tables are broken no idea why... try to fix it.. does not work well...
   # remove caption as this also makes tables unreadable
    contents = re.sub(r'<caption style=.*</caption>', '', contents)
    
    contents = contents.replace("<br/>\n", " ")
    contents = contents.replace("<br/>", " ")    
    contents = contents.replace("\n </td>", "</td>")  
    contents = contents.replace("\n</td>", "</td>")
    contents = contents.replace("${", "``${")
    contents = contents.replace("}", "}``")
    soup2 = BeautifulSoup(contents, SOUP_PARSER)
    
    return md(soup2)
`

then the return value is added to the markdownfile using new_paragraph..

**Comment by didix21** [comment]
Could @osmanburucu-ibm  check if it's fixed on v1.4.0?
